### PR TITLE
Refactor script to use pathlib

### PR DESCRIPTION
**Pull request by ju-sh**
Refactor asciidoc2html script

### CODE REVIEW

**Comment by ju-sh** [comment]
Regarding the mypy problem, I just didn't know where and how to place type hints properly. It's not a Travis problem.

Should we be using type annotations only for the arguments of the functions or should we be using it for variables in the function body as well?

And in [`parse_args()`](https://github.com/qutebrowser/qutebrowser/blob/master/scripts/asciidoc2html.py#L275), while adding the `--website` argument, is the `nargs=1` really required? Can it be removed so that we can get it as a single value instead of a list?

**Comment by The-Compiler** [comment]
> Regarding the mypy problem, I just didn't know where and how to place type hints properly. It's not a Travis problem.

I meant your "# mypy gives error here. Not sure why" comment in the PR - I don't see any mypy errors.

> Should we be using type annotations only for the arguments of the functions or should we be using it for variables in the function body as well?

What I usually do is to annotate functions only, and only add annotations for local variables if mypy says it can't infer them. Note that we'll need to use the old-style syntax:

```python
self.value = 42  # type: int
```

rather than the new-style:

```python
self.value: int = 42
```

because at the moment, qutebrowser still supports Python 3.5.



> And in [`parse_args()`](https://github.com/qutebrowser/qutebrowser/blob/master/scripts/asciidoc2html.py#L275), while adding the `--website` argument, is the `nargs=1` really required? Can it be removed so that we can get it as a single value instead of a list?

Hm, probably not. Want to clean that up as part of this PR?

File: scripts/asciidoc2html.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
 # along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
 
 """Generate the html documentation based on the asciidoc files."""
+from typing import List, Tuple
```
Comment (The-Compiler) [inline scripts/asciidoc2html.py:22]: nitpick: Please move this down to the other imports, as it's a Python stdlib import.

File: scripts/asciidoc2html.py [hunk 2]
```diff
@@ -30,19 +31,26 @@
...
     """Abstraction of an asciidoc subprocess."""
 
     FILES = ['faq', 'changelog', 'contributing', 'quickstart', 'userscripts']
 
-    def __init__(self, asciidoc, website):
+    def __init__(self, asciidoc, website) -> None:
+        """
+        asciidoc: Optional[List[str]]
+        website: Optional[List[str]])
+        """
```
Comment (The-Compiler) [inline scripts/asciidoc2html.py:53]: Please use the annotation syntax (`asciidoc: Optional[List[str]]` etc.) instead - that is supported in all Python 3 versions, only annotating variables isn't.

File: scripts/asciidoc2html.py [hunk 3]
```diff
@@ -51,37 +59,35 @@ def __init__(self, asciidoc, website):
...
         else:
             self._build_docs()
             self._copy_images()
 
-    def _build_docs(self):
+    def _build_docs(self) -> None:
         """Render .asciidoc files to .html sites."""
-        files = [('doc/{}.asciidoc'.format(f),
-                  'qutebrowser/html/doc/{}.html'.format(f))
+        files: List[Tuple[pathlib.Path, pathlib.Path]] = [(pathlib.Path('doc/{}.asciidoc'.format(f)),
```
Comment (The-Compiler) [inline scripts/asciidoc2html.py:86]: You'll need to use the comment syntax here to keep compatibility for Python 3.5.

**Comment by ju-sh** [comment]
> I meant your "# mypy gives error here. Not sure why" comment in the PR - I don't see any mypy errors.

Oh, I had missed deleting that. I had tried using type hints first, but kept getting error, so I removed the hints but forgot about the comment.

> because at the moment, qutebrowser still supports Python 3.5.

Okay. :+1: 

> Hm, probably not. Want to clean that up as part of this PR?

Yeah, I'll do that.

File: scripts/asciidoc2html.py [hunk 4]
```diff
@@ -214,13 +208,13 @@ def _build_website(self):
...
-                (os.path.join('doc', 'quickstart.html'), 'quickstart.html')]:
+                ((pathlib.Path('doc') / 'quickstart.html'), 'quickstart.html')]:
             try:
-                os.symlink(dst, os.path.join(outdir, link_name))
+                (outdir / link_name).symlink_to(dst)    # mypy gives error here. Not sure why
             except FileExistsError:
                 pass
 
-    def _get_asciidoc_cmd(self):
+    def _get_asciidoc_cmd(self):    # -> List[str]
```
Comment (The-Compiler) [inline scripts/asciidoc2html.py:217]:
```suggestion
    def _get_asciidoc_cmd(self) -> List[str]:
```

**Comment by ju-sh** [comment]
What could be the type annotation for [`_homedir`](https://github.com/qutebrowser/qutebrowser/blob/master/scripts/asciidoc2html.py#L49) of `AsciiDoc`?

I figured it to be  `Optional[pathlib.Path]` as it is `None` in the beginning and `pathlib.Path` later. But if I use it, mypy will complain as `/` operator is used on `_homedir` later on.

Any idea how to get around this?

**Comment by The-Compiler** [comment]
There are usually three ways around this kind of thing:

1) Find a way to set it to a proper value without having to set it to `None` first.
2) Use `typing.cast(pathlib.Path, None)` to pretend `None` is actually a path
3) Use something like `assert self._homedir is not None` before using it, which mypy interprets as "we're sure it's set at this point".

Here I'd probably go with 3).

File: scripts/asciidoc2html.py (hunk 1 above)
Comment (ju-sh) [reply scripts/asciidoc2html.py:22]: You mean skip the blank line between this and the other stdlib imports?

Comment (The-Compiler) [reply scripts/asciidoc2html.py:22]: Yeah (and keep the blank line between the docstring above and the imports)

**Comment by ju-sh** [comment]
No idea what happened there... Let me try to fix it.

**Comment by ju-sh** [comment]
Had accidently overwritten a variable... It's working now.

**Comment by The-Compiler** [comment]
Looks good now, thanks! :+1:

**Pull request merged**